### PR TITLE
*: begin a transaction in PrepareTxnCtx, unify in transaction.

### PR DESCRIPTION
**Pull request by coocood**
Transaction should has the same life cycle as TxnCtx, so create a
new transaction in PrepareTxnCtx if transaction is nil or invalid.

Unify ShouldAutocommit to InTxn flag, remove IsDDL method.

### CODE REVIEW

File: session_test.go [hunk 1]
```diff
@@ -310,7 +310,7 @@ func (s *testSessionSuite) TestAutocommit(c *C) {
 
 	checkTxn(c, se, "create table t (id BIGINT PRIMARY KEY AUTO_INCREMENT NOT NULL)", 0)
 	checkAutocommit(c, se, 2)
-	checkTxn(c, se, "set autocommit=0;", 0)
+	checkTxn(c, se, "set autocommit=0;", 1)
```
Comment (zimulala) [inline session_test.go]: Why change this test.

File: executor/executor_simple.go [hunk 2]
```diff
@@ -124,9 +124,8 @@ func (e *SimpleExec) executeBegin(s *ast.BeginStmt) error {
 }
 
 func (e *SimpleExec) executeCommit(s *ast.CommitStmt) error {
-	err := e.ctx.CommitTxn()
 	e.ctx.GetSessionVars().SetStatusFlag(mysql.ServerStatusInTrans, false)
-	return errors.Trace(err)
+	return nil
```
Comment (hanfei1991) [inline executor/executor_simple.go]: if the err is always nil, why return an error?

File: tidb.go [hunk 3]
```diff
@@ -153,33 +157,40 @@ func Compile(ctx context.Context, rawStmt ast.StmtNode) (ast.Statement, error) {
...
-		err = ctx.CommitTxn()
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
-	}
 	rs, err = s.Exec(ctx)
 	// All the history should be added here.
 	se := ctx.(*session)
 	getHistory(ctx).add(0, s)
 	// MySQL DDL should be auto-commit.
```
Comment (hanfei1991) [inline tidb.go]: This comment should be updated?

File: session_test.go (hunk 1 above)
Comment (zimulala) [inline session_test.go]: `set autocommit=0;` shouldn't affect the beginning or committing of a transaction.

File: executor/executor_ddl.go [hunk 4]
```diff
@@ -72,6 +72,7 @@ func (e *DDLExec) Next() (*Row, error) {
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
+	e.ctx.GetSessionVars().SetStatusFlag(mysql.ServerStatusInTrans, false)
```
Comment (zimulala) [inline executor/executor_ddl.go]: Add comments.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**